File: 노영주/45주차/가장많이받은선물.js
function solution(friends, gifts) {
  let answer = 0;
  const giftHistoryMap = new Map();
  const giftFactor = {};

  const getFromToHistory = (from, to) => {
    return `${from} ${to}`;
  };

  const isAGreaterThanB = (A, B, factor) => {
    const pointA = factor[A].sent - factor[A].received;
    const pointB = factor[B].sent - factor[B].received;

    if (pointA > pointB) {
      return true;
    }

    return false;
  };

  friends.forEach(friend => (giftFactor[friend] = { sent: 0, received: 0 }));

  gifts.forEach(history => {
    const [from, to] = history.split(' ');
    if (giftHistoryMap.has(history)) {
      giftHistoryMap.set(history, giftHistoryMap.get(history) + 1);
    } else {
      giftHistoryMap.set(history, 1);
    }

    giftFactor[from].sent++;
    giftFactor[to].received++;
  });

  for (let i = 0; i < friends.length; i++) {
    const friend = friends[i];
    let nextMonthGiftCount = 0;

    for (let j = 0; j < friends.length; j++) {
      if (i === j) {
        continue;
      }
      const compareFriend = friends[j];
      const sentCount = giftHistoryMap.get(
        getFromToHistory(friend, compareFriend),
      );
      const receivedCount = giftHistoryMap.get(
        getFromToHistory(compareFriend, friend),
      );

      if (!sentCount && !receivedCount) {
        if (isAGreaterThanB(friend, compareFriend, giftFactor)) {
          nextMonthGiftCount++;
        }
        continue;
      }

      if (sentCount === receivedCount) {
        if (isAGreaterThanB(friend, compareFriend, giftFactor)) {
          nextMonthGiftCount++;
        }
        continue;
      }

      if (sentCount) {
        if (!receivedCount || (receivedCount && sentCount > receivedCount)) {
          nextMonthGiftCount++;
        }
      }
    }
    answer = Math.max(answer, nextMonthGiftCount);
  }

  return answer;
}

File: 윤호선/45주차/가장많이받은선물.java
import java.util.HashMap;

public class 가장많이받은선물 {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int friendsNum = friends.length;
        int[][] table = new int[friendsNum + 1][friendsNum + 1];
        int[] nextGifts = new int[friendsNum];
        HashMap<String, Integer> friendIdx = new HashMap<>();

        for (int i = 0; i < friends.length; i++) {
            friendIdx.put(friends[i], i);
        }
        for (int i = 0; i < gifts.length; i++) {
            int A = friendIdx.get(gifts[i].split(" ")[0]);
            int B = friendIdx.get(gifts[i].split(" ")[1]);
            table[A][B]++;
            table[A][friendsNum]++;
            table[friendsNum][B]++;
        }
        for (int i = 0; i < friendsNum; i++) {
            for (int j = i + 1; j < friendsNum; j++) {
                if (table[i][j] > table[j][i]) {
                    nextGifts[i]++;
                } else if (table[i][j] < table[j][i]) {
                    nextGifts[j]++;
                } else {
                    int rateA = table[i][friendsNum] - table[friendsNum][i];
                    int rateB = table[j][friendsNum] - table[friendsNum][j];
                    if (rateA > rateB) nextGifts[i]++;
                    else if (rateA < rateB) nextGifts[j]++;
                }
            }
        }
        for (int i = 0; i < nextGifts.length; i++) {
            answer = Math.max(answer, nextGifts[i]);
        }
        return answer;
    }
}
